flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/CalcUpsertToRetractionTransposeRule.scala

```diff
@@ -0,0 +1,109 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.rules.logical
+
+import org.apache.calcite.plan.RelOptRule.{none, operand}
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
+import org.apache.calcite.rex.{RexCall, RexInputRef}
+import org.apache.calcite.sql.SqlKind
+import org.apache.flink.table.plan.nodes.logical.{FlinkLogicalCalc, FlinkLogicalUpsertToRetraction}
+
+import scala.collection.JavaConversions._
+
+/**
+  * Use this rule to transpose Calc through UpsertToRetraction relnode. It is beneficial if we get
+  * smaller state size in upsertToRetraction.
+  */
+class CalcUpsertToRetractionTransposeRule extends RelOptRule(
+  operand(classOf[FlinkLogicalCalc],
+    operand(classOf[FlinkLogicalUpsertToRetraction], none)),
+  "CalcUpsertToRetractionTransposeRule") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val calc = call.rel(0).asInstanceOf[FlinkLogicalCalc]
+    val upsertToRetraction = call.rel(1).asInstanceOf[FlinkLogicalUpsertToRetraction]
+
+    // column pruning or push Filter down
+    calc.getRowType.getFieldCount <= upsertToRetraction.getRowType.getFieldCount &&
+    // key fields should not be changed
+      fieldsRemainAfterCalc(upsertToRetraction.keyNames, calc)
+  }
+
+  override def onMatch(call: RelOptRuleCall) {
+    val calc = call.rel(0).asInstanceOf[FlinkLogicalCalc]
+    val upsertToRetraction = call.rel(1).asInstanceOf[FlinkLogicalUpsertToRetraction]
+
+    // get new calc
+    val newCalc = calc.copy(calc.getTraitSet, upsertToRetraction.getInput, calc.getProgram)
+    // get new upsertToRetraction
+    val oldKeyNames = upsertToRetraction.keyNames
+    val newKeyNames = getNamesAfterCalc(oldKeyNames, calc)
+
+    val newUpsertToRetraction = new FlinkLogicalUpsertToRetraction(
+      upsertToRetraction.getCluster,
+      upsertToRetraction.getTraitSet,
+      newCalc,
+      newKeyNames)
+
+    call.transformTo(newUpsertToRetraction)
+  }
+
+  private def fieldsRemainAfterCalc(fields: Seq[String], calc: FlinkLogicalCalc): Boolean = {
+    // get input output names
+    val inOutNames = getInOutNames(calc)
+    // contains all fields
+    inOutNames.map(_._1).containsAll(fields)
+  }
+
+  /**
+    * Get related output field names for input field names for Calc.
+    */
+  private def getInOutNames(calc: FlinkLogicalCalc): Seq[(String, String)] = {
+    val inNames = calc.getInput.getRowType.getFieldNames
+    calc.getProgram.getNamedProjects
+      .map(p => {
+        calc.getProgram.expandLocalRef(p.left) match {
+          // output field is forwarded input field
+          case r: RexInputRef => (r.getIndex, p.right)
+          // output field is renamed input field
+          case a: RexCall if a.getKind.equals(SqlKind.AS) =>
+            a.getOperands.get(0) match {
+              case ref: RexInputRef =>
+                (ref.getIndex, p.right)
+              case _ =>
+                (-1, p.right)
```

Comment:
Why do you create a tuples `(int, string)` and later filtering out based on integers instead of for example inserting or not something to an array or a map? Another remark, I think that maybe implementing `RexVisitorImpl` (like `InputRefVisitor`) would be a better, more re-usable approach which would for example solve the problem of duplicating the code for `case r: RexInputRef`.

Reply:
Maybe we can use `RexVisitorImpl` in  `FlinkLogicalCalc`?

Reply:
What do you mean? 

Maybe to rephrase my comment: I think that instead of having this `match case` construct a better idea might to implement it's logic as `RexVisitorImpl`, probably as a private static nested class inside `FlinkLogicalCalc` or `CommonCalc`.